include/os/linux/kernel/linux/simd_aarch64.h

```diff
@@ -31,24 +34,83 @@
  *	kfpu_end()
  *	kfpu_init()
  *	kfpu_fini()
+ *
+ * SIMD support:
+ *
+ * Following functions should be called to determine whether CPU feature
+ * is supported. All functions are usable in kernel and user space.
+ * If a SIMD algorithm is using more than one instruction set
+ * all relevant feature test functions should be called.
+ *
+ * Supported features:
+ *   zfs_neon_available()
+ *   zfs_sha256_available()
+ *   zfs_sha512_available()
  */
 
 #ifndef _LINUX_SIMD_AARCH64_H
 #define	_LINUX_SIMD_AARCH64_H
 
-#include <sys/isa_defs.h>
-
-#if defined(__aarch64__)
-
 #include <sys/types.h>
 #include <asm/neon.h>
+#include <asm/elf.h>
+#include <asm/hwcap.h>
+#include <linux/version.h>
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 17, 0)
+#include <asm/sysreg.h>
+#else
+#define	sys_reg(op0, op1, crn, crm, op2) ( \
+	((op0) << Op0_shift) | \
+	((op1) << Op1_shift) | \
+	((crn) << CRn_shift) | \
+	((crm) << CRm_shift) | \
+	((op2) << Op2_shift))
+#endif
+
+#define	ID_AA64PFR0_EL1		sys_reg(3, 0, 0, 1, 0)
+#define	ID_AA64ISAR0_EL1	sys_reg(3, 0, 0, 6, 0)
 
 #define	kfpu_allowed()		1
 #define	kfpu_begin()		kernel_neon_begin()
 #define	kfpu_end()		kernel_neon_end()
-#define	kfpu_init()		0
-#define	kfpu_fini()		((void) 0)
+#define	kfpu_init()		(0)
+#define	kfpu_fini()		do {} while (0)
+
+#define	get_ftr(id) {				\
+	unsigned long __val;			\
+	asm("mrs %0, "#id : "=r" (__val));	\
+	__val;					\
+}
 
-#endif /* __aarch64__ */
+/*
+ * Check if NEON is available
+ */
+static inline boolean_t
+zfs_neon_available(void)
+{
+	unsigned long ftr = ((get_ftr(ID_AA64PFR0_EL1)) >> 16) & 0xf;
+	return (ftr == 0 || ftr == 1);
+}
+
+/*
+ * Check if SHA256 is available
+ */
+static inline boolean_t
+zfs_sha256_available(void)
+{
+	unsigned long ftr = ((get_ftr(ID_AA64ISAR0_EL1)) >> 12) & 0x3;
+	return (ftr & 0x1);
```

Comment:
It might be different on non-Apple arm64, but my M2 reports:
```
<zfs`spl_cpuid_id_aa64isar0_el1 (spl-processor.c:172)> cpu_features0: 0x0221100110212120
<zfs`spl_cpuid_id_aa64isar0_el1 (spl-processor.c:172)> cpu_features0: AES PMULL SHA1 SHA256 SHA512 SHA3
```
Note that the SHA bits  [12-15] (0xf) has the value `0x2`. Which would in the above "if" only say yes to SHA512, but not SHA256.  I take it to mean that if SHA512 is enabled, then SHA256 is implied as as well.  I would think if should
be `if (ftr & 0xf)` being anything not-zero. (or 0x3 if preferred, but either bit)
Is this an Apple quirk?

It also happens with the AES bits [4-7]
```
            aes & 3 ? "AES " : "",
            aes & 2 ? "PMULL " : "",
```
If either bit is set, it has AES, if 0x2 is set, it also has PMULL.